src/compiler/checker.ts

```diff
@@ -10096,21 +10096,29 @@ namespace ts {
             inferTypes(context.signature.typeParameters, context.inferences, originalSource, originalTarget);
         }
 
+        function getSymbolForInference(type: Type) {
+            // Exclude the static side of classes since it shares its symbol with the instance side which leads
+            // to false positives.
+            return type.flags & TypeFlags.Object && !(getObjectFlags(type) & ObjectFlags.Anonymous && type.symbol && type.symbol.flags & SymbolFlags.Class) ? type.symbol : undefined;
+        }
+
         function inferTypes(typeVariables: TypeVariable[], typeInferences: TypeInferences[], originalSource: Type, originalTarget: Type) {
-            let sourceStack: Type[];
-            let targetStack: Type[];
+            let stack: Type[];
             let depth = 0;
             let inferiority = 0;
             const visited = createMap<boolean>();
             inferFromTypes(originalSource, originalTarget);
 
-            function isInProcess(source: Type, target: Type) {
-                for (let i = 0; i < depth; i++) {
-                    if (source === sourceStack[i] && target === targetStack[i]) {
-                        return true;
+            function isInstantiationInProcess(type: Type) {
+                const symbol = getSymbolForInference(type);
+                if (symbol) {
+                    for (let i = 0; i < depth; i++) {
+                        const t = stack[i];
+                        if (getSymbolForInference(t) === symbol) {

```

Comment:
why not storing `getSymbolForInference(target)` on the stack directly instead of calling it on every entry when we check